### ISSUE TITLE
gnu.org was briefly down; how to avoid loading assets like CSS from a remote site

### ISSUE THREAD

**Issue by ernstki** [benzado/gnu-make-dash-docset#4]
It _seems_ like @lshprung has taken over maintenance of this docset, but he doesn't have an issue tracker in his fork (yet?), hence why I'm opening this here. I'm willing to do a PR, wherever that PR needs to land.

gnu.org was briefly down early this morning, US Eastern time. Or rather it wasn't _down_, but the response time was extreeeeeemely long. Perhaps it was just something routine. However, the HTML sources from the docset reference a CSS stylesheet from gnu.org…

![Screenshot of source view for the GNU Make "Automatic Variables" section, showing the reference to manual.css on gnu.org](https://github.com/benzado/gnu-make-dash-docset/assets/4009681/3bcc6663-820d-4faf-bfce-7bebb9afd91e)

…and with the long response time, it made it seem as if Zeal was hung, or the docset was blank. Unless you waited a few minutes for the request to time out, that is.

Just based on some context [here](https://github.com/zealdocs/zeal/issues/955#issuecomment-425327306), it looks like Dash optionally blocks HTTP requests if a docset has some particular key in the `Info.plist` set. But I'm not sure how that `Info.plist` is generated exactly, or if Zeal even supports that setting in `Info.plist` (yet).

In any event, the GNU Make docs _look better_ with the appropriate CSS. So maybe the thing to do in this particular case is to inline it for every page, since [it's quite small](https://www.gnu.org/software/gnulib/manual.css) anyway, or download the `manual.css` and alter all the stylesheet references to be a path relative to the docset.

Do you have any insights on this? What would you recommend?

**Comment by lshprung** [comment]
Issues are now set up on <https://github.com/lshprung/gnu-make-dash-docset>, thanks for bringing it to my attention. Let me do some testing and get back to you. In the meantime, what version of Zeal are you using? The comment [here](https://github.com/zealdocs/zeal/issues/955#issuecomment-1509520413) sounds like Zeal's upstream may have already resolved issues like this (if I am reading the thread correctly).

**Comment by lshprung** [comment]
In any case, I don't like the idea of including a relative path CSS file. Having the stylesheet point to the upstream manual.css file hosted by the GNU project allows us to immediately receive any upstream changes without having to repackage the docset. It sounds like this issue may be better directed towards [Zeal](https://github.com/zealdocs/zeal)/[Dash](https://github.com/Kapeli/Dash-User-Contributions); let me know if you disagree.

**Comment by benzado** [comment]
For what it's worth, my vote would be to embed the CSS file or strip out the `<link>` tag so that, either way, the docset doesn't try to load anything over the network.

Having access to documentation offline is a major feature of both Dash and Zeal, according to their respective home pages:

> Dash gives your Mac instant offline access to 200+ API documentation sets.

and

> Zeal is an offline documentation browser for software developers.

Therefore, taking 30 seconds to render a page from a docset, simply because gnu.org is under the weather, is a bug.

Is the bug in the docset because it has a remote reference? Or is the bug in Dash/Zeal because it doesn't block such references? That's mostly a political question. But it looks like zealdocs/zeal#955 has been open for a quite a long time and any solutions have been docset-specific.

(Also, getting upstream changes to the stylesheets instantly, but changes to the documents with a new docset, is not a guaranteed benefit, because those two parts depend on each other. An updated stylesheet is just as likely to make older documents look worse as better.)

**Comment by lshprung** [comment]
Fair enough. In that case, I can propose for this script to download the upstream CSS file alongside the upstream docs from gnu.org and replace the `<link>` with a relative path for all pages. I can also include a `--no-css` (or something) flag for the Makefile to just remove the links and have the docset be without any CSS. If that sounds good, I can put together a PR on my fork when I have time.

**Comment by ernstki** [comment]
Just another interesting data point: on my Mac, I had an old `~/Library/Application Support/Dash/User Contributed/GNU Make/GNU_Make.docset` from January 2020, and it had _no_ stylesheet at all. So, I guess the change was on gnu.org's end.

![Screenshot of an old (ca 2020) Dash docset for GNU Make, showing where the <style> tag *would* be, if it were there](https://github.com/benzado/gnu-make-dash-docset/assets/4009681/2e2fab62-ed2a-48a3-9126-0bc64ec58dcc)

> Is the bug in the docset because it has a remote reference? Or is the bug in Dash/Zeal because it doesn't block such references? That's mostly a political question. But it looks like [zealdocs/zeal#955](https://github.com/zealdocs/zeal/issues/955) has been open for a quite a long time and any solutions have been docset-specific.

From my perspective, that the docset didn't bundle an offline copy of a remote asset. But at the precise moment I opened the ticket, I also didn't know that there was a way to tell a docset (in its `Info.plist`) not to load remote resources.

In the sense that I offered to do a PR, it's shell scripts and Makefiles and HTML/CSS stuff that I understand and could help out with; C++ and QtWebKit stuff, not as much. `;)`

**Comment by lshprung** [comment]
[PR is up on my branch](https://github.com/lshprung/gnu-make-dash-docset/pull/1). I'm not sure how to add both of you as reviewers, but please take a look.

**Comment by ernstki** [comment]
> In the meantime, what version of Zeal are you using?

@lshprung Overlooked this question originally, my bad. I'm using `0.6.1-1build1` from the Ubuntu 20.04 LTS repositories. The distro is elementary OS 6.1 (Jólnir).

Perhaps it's a bit lazy of me to file the bug without trying the newest upstream version of Zeal, but 0.6.1 _is_ technically the [latest release](https://github.com/zealdocs/zeal/releases).

Silly question: would you like comments on the PR here (in the issue) or actually _on_ the PR?

**Comment by lshprung** [comment]
I suppose it's easier to comment here. Was there any feedback you had?

**Comment by lshprung** [comment]
I merged my branch and will create a pull request for Dash-User-Contributions shortly. The issue can be closed

**Comment by ernstki** [comment]
> I suppose it's easier to comment here. Was there any feedback you had?

I think I may've found a bug in `src/set-stylesheet.sh` when it didn't work out of the box for me. I started to look into it, but had to put it aside for the moment.

If _you_ are able to build the docset, it's not an issue that needs to hold up the show. I'll open a PR somewhere—here? or your fork?—if and when I get to the bottom of that.

![Screenshot of a possible bug in 'src/set-stylesheet.sh'](https://github.com/benzado/gnu-make-dash-docset/assets/4009681/d12f7a7a-66dd-4abf-aee1-a230f24b4ded)

**Comment by lshprung** [comment]
> > I suppose it's easier to comment here. Was there any feedback you had?
> 
> I think I may've found a bug in `src/set-stylesheet.sh` when it didn't work out of the box for me. I started to look into it, but had to put it aside for the moment.
> 
> If _you_ are able to build the docset, it's not an issue that needs to hold up the show. I'll open a PR somewhere—here? or your fork?—if and when I get to the bottom of that.
> 
> ![Screenshot of a possible bug in 'src/set-stylesheet.sh'](https://user-images.githubusercontent.com/4009681/249663252-d12f7a7a-66dd-4abf-aee1-a230f24b4ded.png)

Ok. Up to you if you want to migrate the issue to my fork or continue to discuss here.

**Comment by benzado** [comment]
> It _seems_ like [@lshprung](https://github.com/lshprung) has taken over maintenance of this docset

It seems that way to me, too! 😆 I should have archived this repo years ago. Finally doing some cleanup now.

I poked around to see what Louie has done, and... wow!

* He rewrote my Ruby scripts as shell scripts (good move, I love Ruby but the installation process is far too temperamental to be a dependency for something like this). lshprung/gnu-make-dash-docset@e397123c70d2ad564753f8aad1e1ac82276234f0
* He set up a meta-project for building this and several other GNU docsets: https://github.com/lshprung/gnu-make-dash-docset
* He knows how to use git submodules! lshprung/dash-docset-builder@6b2603325822049f31f003781aa9761799aa4ce7
* He rewrote the scripts again in Python!  lshprung/gnu-make-dash-docset@0dc6b48d5a99d7dff2aa419c7bfbe1caadcb8944
* In spite of it being a literal [Ship of Theseus](https://en.wikipedia.org/wiki/Ship_of_Theseus), he gave me credit 😊 lshprung/dash-docset-builder@a26adab1566f7ad0574c93b841f640363a799092

The decision to share your source code is one that has unfortunately become more complicated with each passing year. I made a docset mainly because I _didn't_ use GNU Make very often, and so I was always annoyed when I couldn't find the docs in Dash. Arguably, I shouldn't have published it, because I had no plans to maintain it. But I'm glad I did. I don't know that @lshprung really needed my script as a starting point, except maybe as motivation per [Cunningham's Law](https://meta.wikimedia.org/wiki/Cunningham's_Law). Seeing how it has transformed it into a thing of beauty warms my heart.

Sorry it took me so long to close this issue!

Fixed by lshprung/gnu-make-dash-docset#1

Released by Kapeli/Dash-User-Contributions#4370
